### ISSUE TITLE
VCF comparison - TypeError: '>=' not supported between instances of 'int' and 'NoneType'

### ISSUE THREAD

**Issue by josealvesferrao** [spiralgenetics/truvari#96]
Hello everyone,

We´re using Truvari (v3.0.1) bench to compare VCF files from ONT dataset (using cuteSV) vs microarray dataset (vcf exported from ChAS-Chromosome Analysis Suite software). When comparing two vcfs only from ONT, everything works fine, but when we use a vcf exported from ChAS (vcf version 4.2, the same as the one from ONT) we get a Python error. All vcfs are sorted, gzipped and indexed.

Command: truvari bench -b 123456_CytoScanHD_Array.cyhd.ND.sorted.vcf.gz -c /full/path/123456_cutesv_filtered.vcf.gz -f /full/path/Hg38/Homo_sapiens.GRCh38.dna.primary_assembly.fa -o /full/path/Truvari1/

Error:
Traceback (most recent call last):
  File "/usr/local/bin/truvari", line 8, in <module>
    sys.exit(main())
  File "/usr/local/lib/python3.10/site-packages/truvari/__main__.py", line 64, in main
    TOOLS[args.cmd](args.options)
  File "/usr/local/lib/python3.10/site-packages/truvari/bench.py", line 439, in bench_main
    regions = truvari.GenomeTree(outputs["vcf_base"], outputs["vcf_comp"], args.includebed, args.sizemax)
  File "/usr/local/lib/python3.10/site-packages/truvari/genome_tree.py", line 21, in __init__
    self.tree = self.__build_tree(vcfA, vcfB)
  File "/usr/local/lib/python3.10/site-packages/truvari/genome_tree.py", line 52, in __build_tree
    all_regions[name].addi(0, length)
  File "/usr/local/lib/python3.10/site-packages/intervaltree/intervaltree.py", line 343, in addi
    return self.add(Interval(begin, end, data))
  File "/usr/local/lib/python3.10/site-packages/intervaltree/intervaltree.py", line 323, in add
    if interval.is_null():
  File "/usr/local/lib/python3.10/site-packages/intervaltree/interval.py", line 139, in is_null
    return self.begin >= self.end
TypeError: '>=' not supported between instances of 'int' and 'NoneType'

We´re guessing that the error must be related with ChAS vcf file configuration, maybe some important information that Truvari needs is missing or incorrect.

ChAS VCF example:
#CHROM	POS	ID	REF	ALT	QUAL	FILTER	INFO	FORMAT	123456_(CytoScanHD_Array).cyhd.ND.cychp
1	72293841	.	N	<CNV>	.	.	SVTYPE=CNV;END=72346221;SVLEN=52380;IMPRECISE	CN	4
1	173912764	.	N	<CNV>	.	.	SVTYPE=CNV;END=173963653;SVLEN=50889;IMPRECISE	CN	3
10	90317517	.	N	<CNV>	.	.	SVTYPE=CNV;END=90843786;SVLEN=526269;IMPRECISE	CN	3
11	55606559	.	N	<CNV>	.	.	SVTYPE=CNV;END=55693312;SVLEN=86753;IMPRECISE	CN	0
12	45282874	.	N	<CNV>	.	.	SVTYPE=CNV;END=45353285;SVLEN=70411;IMPRECISE	CN	3

ONT VCF example:
#CHROM	POS	ID	REF	ALT	QUAL	FILTER	INFO	FORMAT	EOL-1_21-10-2021
1	181211	cuteSV.INS.3	G	AAATTTGGGCCC	16.9	PASS	PRECISE;SVTYPE=INS;SVLEN=55;END=181211;CIPOS=-36,36;CILEN=-3,3;RE=3;RNAMES=5bdd1234-478a-4e38-8b55-c3eeac2fc3ac,66bd1bb2-2b32-43a7-b90f-d9b568dc1df4,55c31918-de29-4e4c-8a6d-aa9fb6c80306;AF=0.6	GT:DR:DV:PL:GQ	0/1:2:3:17,1,7:6
1	350807	cuteSV.INS.4	A	AAATTTGGGCCCAAA	13.7	PASS	PRECISE;SVTYPE=INS;SVLEN=39;END=350807;CIPOS=0,0;CILEN=0,0;RE=3;RNAMES=f47d2e5a-e321-41e3-8418-ebb2be67e9b3,bdebecd9-6c7e-4d59-a558-2bd2a3456771,a3354bf2-4e36-4131-b821-f98b20f07b47;AF=0.5	GT:DR:DV:PL:GQ	0/1:3:3:14,0,14:10

Why do you think we are getting this error? Can you please help us solve it?

Regards,

Jose

**Comment by ACEnglish** [comment]
Given your example, the problem is likely with the ONT VCF. I believe the [VCF format](https://samtools.github.io/hts-specs/VCFv4.2.pdf) uses 1-based, half opened intervals and insertions/deletions need to have an anchor base. So the line
```
1 181211 cuteSV.INS.3 G AAATTTGGGCCC 16.9 PASS PRECISE;SVTYPE=INS;SVLEN=55;END=181211;CIPOS=-36,36;CILEN=-3,3;RE=3;RNAMES=5bdd1234-478a-4e38-8b55-c3eeac2fc3ac,66bd1bb2-2b32-43a7-b90f-d9b568dc1df4,55c31918-de29-4e4c-8a6d-aa9fb6c80306;AF=0.6 GT:DR:DV:PL:GQ 0/1:2:3:17,1,7:6
```
is seems malformed because `POS == END` and `REF[0] != ALT[0]`. If I had to guess, the VCF entry probably needs to look like one of the following (truncated for readibility)

```
1 181210 cuteSV.INS.3 G GAAATTTGGGCCC 16.9 PASS END=181211 GT 0/1

1 181211 cuteSV.INS.3 A AAATTTGGGCCC 16.9 PASS END=181212 GT 0/1
```

Though I'm also now noticing that `len(ALT) != SVLEN`, which is strange. But that should just cause weirdness in the reporting rather than prevent Truvari from  running.

If you ever suspect a VCF is incorrect, read [the wiki](https://github.com/spiralgenetics/truvari/wiki#before-you-start) for instructions on a way to catch a lot of possible formatting errors.

**Comment by ACEnglish** [comment]
I'm sorry, I jumped right to the VCF entries without carefully reading the error.

What I described above is correct and likely should be fixed, but this specific error is caused by the ChAS VCF's header being malformed. 
The header is probably missing a line such as `##contig=<ID=1,length=248956422>`.  The contig name probably exists, but its almost certainly missing the 'length'.

**Comment by ACEnglish** [comment]
Because the VCF Format explicitly states that the `contig` header is optional, but Truvari requires it when `--includebed` is not provided, I'll consider this a bug and put in a ticket for Truvari to catch when this happens and give a more informative error.

**Comment by josealvesferrao** [comment]
Hi Adam,

Thank you very much for the quick answer. I am very sorry about the ONT vcf file example, I had truncated it, that´s why REF[0] != ALT[0] on the first variant and len(ALT) != SVLEN on both variants, but indeed originally POS == END, maybe due to the fact that these variants are insertions (?), as we checked the following ones (deletions and duplications) and POS != END.

You are right, we checked ChAS vcf and indeed the contig length is missing! We will add the length and retry the comparision.

Thanks!

Regards,
Jose

**Comment by ACEnglish** [comment]
Excellent. Let me know if that doesn't fix the problem.

As for `POS == END`, that probably won't prevent Truvari from running, but it may have implications for measuring sequence similarity (`--seqsim`).  It's very minor, and maybe even speculative, but I know that I built/tested Truvari with insertions that had the anchor-base, so END = POS + 1.  So when `POS == END`, I think the entry should become something like:

```
1 181211 cuteSV.INS.3 . AAATTTGGGCCC END=181211
```
Where there is no sequence inside of the REF column. IDK, VCF can get wonky so watch out. 

Closing this ticket for now and tracking development in a [ticket](https://github.com/spiralgenetics/truvari/projects/1)
